public/app/plugins/datasource/prometheus/datasource.ts

```diff
@@ -300,29 +300,26 @@ export class PrometheusDatasource extends DataSourceWithBackend<PromQuery, PromO
     };
   };
 
-  shouldRunExemplarQuery(target: PromQuery): boolean {
-    /* We want to run exemplar query only for histogram metrics:
-    1. If we haven't processd histogram metrics yet, we need to check if expr includes "_bucket" which means that it is probably histogram metric (can rarely lead to false positive).
-    2. If we have processed histogram metrics, check if it is part of query expr.
-    */
+  shouldRunExemplarQuery(target: PromQuery, request: DataQueryRequest<PromQuery>): boolean {
     if (target.exemplar) {
-      const histogramMetrics = this.languageProvider.histogramMetrics;
-
-      if (histogramMetrics.length > 0) {
-        return !!histogramMetrics.find((metric) => target.expr.includes(metric));
-      } else {
-        return target.expr.includes('_bucket');
+      // We check all already processed targets and only create exemplar target for not used metric names
+      const metricName = this.languageProvider.histogramMetrics.find((m) => target.expr.includes(m));
+      const targets = [...request.targets];
+      // Remove targets that weren't processed yet (in targets array they are after current target)
+      targets.splice(request.targets.findIndex((t) => t.refId === target.refId));
```

Comment:
can that `findIndex` fail? i mean, is it possible that it does not find that item and returns `-1`?

Reply:
maybe we could change the
```js
const targets = [...request.targets]
targets.splice(something)
```

to

```js
const targets = request.targets.slice(0, something);
```
?
what do you think?

Reply:
yeah, good suggestion. fixed in 04a2b49

Reply:
> can that findIndex fail

No, tI don't this there can be a situation when this fails